WordPress/gutenberg
# View: Fix TypeScript types

Required for #60796

## What?

Fixes type errors related to the `View` component that are surfaced when `@types/react` is updated to v18.2.79.

## Why?

`@types/react@18.2.79` is a bit stricter, surfacing a problem in the `View` component type. Or more accurately, a problem with how we use the `WordPressComponentProps` utility type. It just happened to manifest at the View component, when consumer components like `Navigator` [used it polymorphically](https://github.com/WordPress/gutenberg/blob/fadd9e0309a5120b734bfc83d7a58254a8b28146/packages/components/src/navigator/navigator-button/hook.ts#L25) and triggered actual prop discrepancies between `div` and `button`.

This is a minimal illustration of how we need to write our types, for a consumer-provided `as` prop to properly influence the accepted props:

```diff
- function Foo( { as, ...restProps }: WordPressComponentProps< {}, 'div' > ) {
+ function Foo< T extends React.ElementType = 'div' >( {
+  as,
+  ...restProps
+ }: WordPressComponentProps< {}, T > ) {
  const Component = as || 'div';
  return <Component { ...restProps } />;
}

// This errors in the first implementation, but is correct in the second.
// (`href` is not a valid prop for `div`, but valid for `a`)
<Foo as="a" href="https://example.com" />;
```

I believe this kind of problem could manifest in any of our polymorphic components that use `WordPressComponentProps` to define props without having a generic at the component-level.

## How?

Reorganize the View component so it can be typed better, and add a generic to better support polymorphism.

## Testing Instructions

Type checks should pass in the current package.json setup, as well as this one with updated React types:

```diff
diff --git a/package.json b/package.json
index e020a7126b..f98ca35651 100644
--- a/package.json
+++ b/package.json
@@ -137,9 +137,9 @@
 		"@types/npm-package-arg": "6.1.1",
 		"@types/prettier": "2.4.4",
 		"@types/qs": "6.9.7",
-		"@types/react": "18.0.21",
+		"@types/react": "18.2.79",
 		"@types/react-dates": "21.8.3",
-		"@types/react-dom": "18.0.6",
+		"@types/react-dom": "18.2.25",
 		"@types/requestidlecallback": "0.3.4",
 		"@types/semver": "7.3.8",
 		"@types/sprintf-js": "1.1.2",
```
